### DISCUSSION TITLE
Swap LoRA adapters at runtime

### DISCUSSION THREAD

**Discussion by ltoniazzi** [ggml-org/llama.cpp#7850] [Ideas]
## New Feature?

Compile LLMs in gguf so that they take an additional int parameter that allows to swap between different LoRA adapters at runtime.

## Questions 
1. Is this feature possible to implement now with llama.cpp (if not already available)?

2. If yes to 1 (with feature not currently available), how difficult is it to develop this feature?

## Related discussions

A [discussion opened ~1year ago](https://github.com/ggerganov/llama.cpp/discussions/1101#discussion-5115082), with the main approach being optimizing tall-skinny matmul to avoid cacheing LoRA weights in the un-merged PR #996. Has this project progressed in other directions?

## Example

[Basic onnx example here](https://github.com/ltoniazzi/adapter-compiler/blob/main/adapter_compiler/main.py), to clarify what this  feature aims to do.

## Context/Application
I wanted to use this for gaming applications, so that one can fine-tune multiple LoRA-adapters for multiple NPCs so that these NPCs can all benefit from the same large model being loaded in memory only once.

**Comment by FSSRepo** [comment]
That would imply keeping the base model and the fine-tuned model in RAM. When the LoRA is changed, it would involve freeing the fine-tuned model and making a copy of the base model with the LoRA applied.

**Reply by slaren** [reply]
It would be possible to apply the LoRAs at inference time with a small overhead, which would make changing LoRAs instant. We would need to replace the calls to `ggml_mul_mat` with something like this:
```C++
static ggml_tensor * lora_mul_mat(llama_context & lctx, ggml_context * ctx0, ggml_tensor * weight, ggml_tensor * cur) {
    ggml_tensor * mm = ggml_mul_mat(ctx0, weight, cur);

    if (!lctx.lora_loaded) {
        return mm;
    }

    auto it = lctx.lora_weights.find(weight->name);
    if (it == lctx.lora_weights.end()) {
        return mm;
    }

    ggml_tensor * loraA = it->second.loraA;
    ggml_tensor * loraB = it->second.loraB;

    ggml_tensor * t_lora =
        ggml_mul_mat(ctx0,
            loraB,
            ggml_mul_mat(ctx0,
                loraA,
                cur
            )
        );

    if (*(float*)lctx.lora_scale->data != 1.0f) {
        t_lora = ggml_scale(ctx0, t_lora, lctx.lora_scale);
    }

    ggml_tensor * t_patch = ggml_add(ctx0, mm, t_lora);
    return t_patch;
}
```

However there is still the problem of conversion of the PEFT LoRAs, due to the permutations and other transformations that are applied to the models during conversion to gguf, a different conversion script would need to be created for each model architecture (which is why `convert-lora-to-ggml.py` was removed).

**Reply by ltoniazzi** [reply]
Thanks @slaren! I'm happy to take a look at the PEFT conversion. 

Can you point me to:
1. the PR that removed `convert-lora-to-ggml.py` that you mentioned and
2. any reference that might come to mind that gives an idea of how to tackle PEFT conversion for gguf

**Reply by slaren** [reply]
It was removed in https://github.com/ggerganov/llama.cpp/pull/7204. You would need to look at `convert-hf-to-gguf.py` for the transformations that are applied to each architecture and do something similar to the lora tensors. For example, for llama architecture these tensors are permuted:
https://github.com/ggerganov/llama.cpp/blob/fd5ea0f897ecb3659d6c269ef6f3d833e865ead7/convert-hf-to-gguf.py#L1349-L1352

**Reply by ltoniazzi** [reply]
On the compilation side of this problem:
1. Should I write the lora weights in the same gguf file with the base weights? 
2. If yes to (1), are there other scripts in the repo to get inspiration from for how to compile base weights along with lora weights in the same gguf file?
3. If no to (1), what is a good strategy to compile the lora weights?

**Reply by slaren** [reply]
I think it would be better if the lora weights are stored in different files than the models, at least to start. You could use the same lora file format that there is already used for finetune and the old `convert-lora-to-ggml.py`, but the best option would probably be to use gguf.

**Reply by ltoniazzi** [reply]
Yes thanks! That's where I was heading, passing a base gguf model + multiple .bin files for each lora model.

Would it be useful as a first iteration to restrict this functionality to models that can be finetuned with `./finetune`? 
This way one does not have to worry about involving PEFT as the lora adapters are validated from the llama.cpp finetuning pipeline.

**Reply by slaren** [reply]
It could be ok, but it would need changes. `finetune` creates lora adapters for every tensor in the weights. I think that's a mistake, it should create loras for the linear layers used in matrix multiplications only. For this reason I don't think it will be possible to apply the loras created by `finetune` at runtime (without merging them into the model weights).

**Reply by ltoniazzi** [reply]
@slaren I managed to apply the LoRA adapters to `Qcur`, `Kcur`, and `Vcur` by substituting in  `ggml_mul_mat` with `lora_mul_mat` (similar to your suggested one) inside `build_llama`. 
So other non-matrix multiplication layers are not applied to the graph (though they are in `llama_context`).

The code runs on CPU but it errors on M2.

<details>
<summary>Click for error on M2</summary>
<br>

Process to get the error: `make` then run `main` passing `open-llama/ggml-model-q8_0.gguf` and a `.bin` lora file obtained from `finetune`  in [this adjusted code with](https://github.com/ggerganov/llama.cpp/compare/master...ltoniazzi:llama.cpp:first-notes#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR7674) `lora_mul_mat`

File that raises `ggml-metal.m`:
```cpp
static id<MTLBuffer> ggml_metal_get_buffer(struct ggml_tensor * t, size_t * offs) {
    //GGML_METAL_LOG_INFO("%s: data tensor '%16s', offs_data = %8ld, offs_eval = %8ld, offs_cach = %8ld\n", __func__, t->name, offs_data, offs_eval, offs_cach);

    const int64_t tsize = ggml_nbytes(t);

    ggml_backend_buffer_t buffer = t->view_src ? t->view_src->buffer : t->buffer;
     // Errors on this line
    struct ggml_backend_metal_buffer_context * buf_ctx = (struct ggml_backend_metal_buffer_context *) buffer->context;
```
With error in debug mode
```bash
Exception has occurred.
EXC_BAD_ACCESS (code=1, address=0x50)
```
and output while running `main`
```
ggml_metal_init: allocating
ggml_metal_init: found device: Apple M2
ggml_metal_init: picking default device: Apple M2
ggml_metal_init: default.metallib not found, loading from source
ggml_metal_init: GGML_METAL_PATH_RESOURCES = nil
ggml_metal_init: loading '/Users/Lorenzo.Toniazzi/Workspace/forks/llama.cpp/ggml-metal.metal'
ggml_metal_init: GPU name:   Apple M2
ggml_metal_init: GPU family: MTLGPUFamilyApple8  (1008)
ggml_metal_init: GPU family: MTLGPUFamilyCommon3 (3003)
ggml_metal_init: GPU family: MTLGPUFamilyMetal3  (5001)
ggml_metal_init: simdgroup reduction support   = true
ggml_metal_init: simdgroup matrix mul. support = true
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: recommendedMaxWorkingSetSize  = 11453.25 MB
llama_kv_cache_init:      Metal KV buffer size =   650.00 MiB
llama_new_context_with_model: KV self size  =  650.00 MiB, K (f16):  325.00 MiB, V (f16):  325.00 MiB
llama_new_context_with_model:        CPU  output buffer size =     0.12 MiB
ggml_gallocr_reserve_n: reallocating Metal buffer from size 0.00 MiB to 521.88 MiB
ggml_gallocr_reserve_n: reallocating CPU buffer from size 0.00 MiB to 10.26 MiB
llama_new_context_with_model:      Metal compute buffer size =   521.88 MiB
llama_new_context_with_model:        CPU compute buffer size =    10.26 MiB
llama_new_context_with_model: graph nodes  = 1075
llama_new_context_with_model: graph splits = 2
zsh: segmentation fault  ./main -m models/open-llama/ggml-model-q8_0.gguf -hl  -n 128
```

</details>


1. Is this the error you are expecting when you say 
    > don't think it will be possible to apply the loras created by finetune at runtime (without merging them into the model weights).
  
    Can you explain a little what the issue is here?

2. By the way, is there an Issue/PR open for removing non-multiplying lora layers from finetune?

**Reply by slaren** [reply]
This is not related to the issue with finetune. You need to allocate the lora tensors using ggml-backend, and using the same buffer type as the type of the layer of the model, to avoid having to copy the tensors constantly to VRAM. `llama_kv_cache_init` shows how to do this.

The issue with `finetune` is that it creates lora adapters for every tensor, and that includes the token embeddings tensors and the norm tensors. These tensors are not used in matrix multiplications, so the `lora_mul_mat` function won't work for them. It would be possible to create similar functions for other operations, but it may not be very efficient, and I don't think loras were ever intended to be used for these layers.

**Reply by ltoniazzi** [reply]
@slaren Sorry but I'm quite new to this. Can you give a couple more details on the ggml-backend? Mainly:

1. What are the essential building blocks for this allocation (get buffer for a layer, create tensors there, move weights to the buffer, etc)?
2. I tried allocating space for lora tensors in `llama_kv_cache_init` like this:

  ```cpp
for (int i = 0; i < (int) n_layer; i++) {
            struct ggml_context * ctx = offload ? ctx_map.at(model.buft_layer[i].buft) : cache.ctxs.front();
            // Allocate space for every (say 5) lora pair in each layer
            for (int n = 0; n < 5; n++) {
                ggml_tensor * loraA = ggml_new_tensor_1d(ctx, type_k, 512*4);
                ggml_tensor * loraB = ggml_new_tensor_1d(ctx, type_k, 512*4);
                ggml_format_name(loraA, "cache_loraA_l%d%d", i, n);
                ggml_format_name(loraB, "cache_loraB_l%d%d", i, n);
                cache.k_l.push_back(loraA); 
                cache.v_l.push_back(loraB);
            }  
    }
```
   
  but I get the same error as before, and I do not see how I can allocate the existing weights to the respective buffer.

**Reply by slaren** [reply]
Since each layer may be stored in a different device, you have to allocate the lora tensors in the memory of the device in which they are used. Otherwise, it will result in lots of data being copied between devices during inference.

Buffer types are things like "CPU buffer", "CUDA buffer" or "Metal buffer". Basically every backend has a different buffer type that represents the local memory of the device.

The basic process is like this:
- Create a `ggml_context` for each buffer type used in the model
- Create tensors in the context corresponding to their buffer type
- Use `ggml_backend_alloc_ctx_tensors_from_buft` to allocate all the tensors in each of the contexts in a buffer of its corresponding type
- Use `ggml_backend_tensor_set` to load the data from the file into the tensors

Another example of how to do this is in `llama_control_vector_init`. There is one control vector for each layer, and it needs to be allocated in the same device as the layer.

You can simplify all of this for testing, and if you are using CPU only, and just run with `-ngl 0` and allocate everything in a CPU buffer by using `ggml_backend_alloc_ctx_tensors_from_buft(ctx, ggml_backend_cpu_buffer_type());`, or run with `-ngl 99` and use a Metal buffer type instead with `ggml_backend_metal_buffer_type()`. It will still run if you allocate the tensors in the wrong buffer type as long as you use ggml-backend to allocate the tensors, it will just be slower.

**Reply by ltoniazzi** [reply]
@slaren I managed to run allocating lora tensors in a Metal buffer, by doing context creation + tensors creation + buffer allocation + data allocation [like this](https://github.com/ggerganov/llama.cpp/pull/8056/files#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR252):

<details>
<summary>load and allocate lora in a context</summary>
<br>

```c
struct lora_data * result = new struct lora_data;
    result->ctx = NULL;
    ...
    struct ggml_init_params params_ggml;
    params_ggml.mem_size   = ggml_tensor_overhead() * GGML_DEFAULT_GRAPH_SIZE;
    params_ggml.mem_buffer = NULL;
    params_ggml.no_alloc   = true;
    result->ctx = ggml_init(params_ggml);
    ...
    while(!file.eof()) {
        ...
        struct ggml_tensor * tensor = ggml_new_tensor(result->ctx, (enum ggml_type) type, n_dims, ne);
        ggml_set_name(tensor, name_buf.data());
        ...
        file.seek(nbytes, SEEK_CUR);
    }

    ggml_backend_buffer_t buf = ggml_backend_alloc_ctx_tensors_from_buft(result->ctx,  ggml_backend_metal_buffer_type());
    ...
    // read tensor data
    result->data.resize(total_nbytes_pad);
    size_t data_offset = 0;
    for (size_t i = 0; i < tensors.size(); ++i) {
        struct ggml_tensor * tensor = tensors[i];
        ...
        tensor->data = result->data.data() + data_offset;
        file.read_raw(tensor->data, nbytes);
        data_offset += nbytes_pad;
    }
    return result;
```

</details>


So the code runs on Metal (and I can debug seeing the lora tensors in a metal buffer), but it is incredibly slow and shows the `GGML_METAL_LOG_ERROR`of the `ggml_metal_get_buffer` function after each generated token for all lora tensors, such as 
```bash
ggml_metal_get_buffer: error: tensor 'blk.23.attn_v.weight.loraA' buffer is nil
```

The methods raising is this in `ggml-metal.m`:

<details>
<summary>ggml_metal_get_buffer</summary>
<br>


```c
static id<MTLBuffer> ggml_metal_get_buffer(struct ggml_tensor * t, size_t * offs) {
    //GGML_METAL_LOG_INFO("%s: data tensor '%16s', offs_data = %8ld, offs_eval = %8ld, offs_cach = %8ld\n", __func__, t->name, offs_data, offs_eval, offs_cach);

    const int64_t tsize = ggml_nbytes(t);

    ggml_backend_buffer_t buffer = t->view_src ? t->view_src->buffer : t->buffer;

    struct ggml_backend_metal_buffer_context * buf_ctx = (struct ggml_backend_metal_buffer_context *) buffer->context;

    // find the view that contains the tensor fully
    for (int i = 0; i < buf_ctx->n_buffers; ++i) {
        const int64_t ioffs = (int64_t) t->data - (int64_t) buf_ctx->buffers[i].data;

        //GGML_METAL_LOG_INFO("ioffs = %10ld, tsize = %10ld, sum = %10ld, buf_ctx->buffers[%d].size = %10ld\n", ioffs, tsize, ioffs + tsize, i, buf_ctx->buffers[i].size);
        if (ioffs >= 0 && ioffs + tsize <= (int64_t) buf_ctx->buffers[i].size) {
            *offs = (size_t) ioffs;

            //GGML_METAL_LOG_INFO("%s: tensor '%16s', offs = %8ld\n", __func__, t->name, *offs);

            return buf_ctx->buffers[i].metal;
        }
    }

    GGML_METAL_LOG_ERROR("%s: error: tensor '%s' buffer is nil\n", __func__, t->name);

    return nil;
}
```

</details>

Can you point me to what the issue is? 
Is the issue that I should put the lora tensors in the exact same context where the [corresponding base tensors for the `mul_mat`s are located](https://github.com/ggerganov/llama.cpp/blob/master/src/llama.cpp#L5696)?

**Reply by slaren** [reply]
I have added a review comment where I think is the problem. To understand how ggml-backend is used, it may help you to check the simple and gpt2 examples in the ggml repository:
https://github.com/ggerganov/ggml/blob/master/examples/simple/simple-backend.cpp
https://github.com/ggerganov/ggml/blob/master/examples/gpt-2/main-backend.cpp

**Comment by MaTwickenham** [comment]
@ltoniazzi Hi there, I am also paying attention to adapters Swap. I saw that you have implemented this on the CPU. Could you provide relevant information on the use of it, such as the configuration and loading of the LoRA model. Thank you very much for your time and assistance!

**Reply by ltoniazzi** [reply]
@MaTwickenham No problem, it's all in a messy branch in my fork now ([branch first-notes](https://github.com/ltoniazzi/llama.cpp/tree/first-notes)). 

Main logic:
1. Created a lora adapter using the `./finetune` example. 
2. This adapter's `.bin` file is hardcoded in the llama.cpp file in `lora.filename = "./models/open-llama/lora-ggml-model-q8_0-shakespeare-LATEST.bin";`
3. Running main `.main -m ./models/open-llama/ggml-model-q8_0.gguf -n 128 -ngl 0` the adapters are loaded in `llama_new_context_with_model` into a [new mapping `lora_weight_map`](https://github.com/ggerganov/llama.cpp/compare/master...ltoniazzi:llama.cpp:first-notes#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR2518) in `llama_context`, and these mappings are picked up by the [new function `lora_mul_mat`](https://github.com/ggerganov/llama.cpp/compare/master...ltoniazzi:llama.cpp:first-notes#diff-150dc86746a90bad4fc2c3334aeb9b5887b3adad3cc1459446717638605348efR9697), doing basically what @slaren suggested [here](https://github.com/ggerganov/llama.cpp/discussions/7850#discussioncomment-9729317).

Before cleaning my code I wanted to make the metal implementation run but I haven't had time this week to look at it.
But happy to clean up the code later today if it helps you, let me know!

**Reply by MaTwickenham** [reply]
@ltoniazzi Really appreciate that! It will help a lot. I have been doing work related to adapter swap recently. Let’s keep in communication.😃

**Reply by ltoniazzi** [reply]
@MaTwickenham I opened the draft PR #8056. The PR description explains how to create the adapter and how to run with the adapter loaded. Let me know how it goes 😃

**Reply by MaTwickenham** [reply]
@ltoniazzi Very helpful, thank you! I am trying to create a vector in the context variable to store multiple adapters based on your code, in order to test the overhead of adapters hot swapping. Additionally, do you have any social media? I would like to have some real-time communication with you. If it's convenient, could you leave your contact information? BTW, My homepage has my email address.
